### PR TITLE
[CCR] Make cross cluster replication work with security

### PR DESCRIPTION
**Pull request by martijnvg**
If security is enabled today with ccr then the follow index api will
fail with the fact that system user does not have privileges to use
the shard changes api. The reason that system user is used is because
the persistent tasks that keep the shards in sync runs in the background
and the user that invokes the follow index api only start those background
processes.

I think it is better that the system user isn't used by the persistent
tasks that keep shards in sync, but rather runs as the same user that
invoked the follow index api and use the permissions that that user has.
This is what this PR does, and this is done by keeping track of
security headers inside  the persistent task (similar to how rollup does this).

This PR also adds a cluster ccr privilege that allows a user to follow
or unfollow an index. Additionally if a user that wants to follow an index,
it needs to have read and monitor privileges on the leader index and
monitor and write privileges on the follow index.

### CODE REVIEW

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java
Comment (jaymode) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java]: can we avoid this? ML and Watcher needed this for BWC since watches/jobs could have existed without the user information. For users that move from a unsecured cluster to a secured cluster, they would just need to update the following of an index?

I should have said the same thing for rollup when I reviewed that

**Review by jaymode — COMMENTED** [review]
The approach makes sense to me, I left a comment about one aspect that I do not think we should do.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java
Comment (martijnvg) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java]: I wasn't entirely sure what the purpose of this was. I saw it being used in all other xpack features, so I just used it here too. I will remove it. 

So if someone is following an index without security enabled and likes to enable security then the user would need to unfollow the index, and then follow the index again (this time the security headers would be automatically applied, assuming that the has been authenticated correctly).

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java
Comment (jasontedor) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java]: I think we want to address this with:

```
compileJava.options.compilerArgs << "-Xlint:-try"
```

 in the CCR `build.gradle`.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java
Comment (jasontedor) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java]: This is really difficult to read. How about:

```diff
diff --git a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java
index 37f248040ee..084b37eaa12 100644
--- a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java
+++ b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java
@@ -385,9 +385,11 @@ public class ShardFollowTasksExecutor extends PersistentTasksExecutor<ShardFollo
                 .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
             return new FilterClient(client) {
                 @Override
-                protected <Request extends ActionRequest, Response extends ActionResponse, RequestBuilder extends
-                    ActionRequestBuilder<Request, Response, RequestBuilder>> void doExecute(Action<Request, Response,
-                    RequestBuilder> action, Request request, ActionListener<Response> listener) {
+                protected <
+                        Request extends ActionRequest,
+                        Response extends ActionResponse,
+                        RequestBuilder extends ActionRequestBuilder<Request, Response, RequestBuilder>>
+                void doExecute(Action<Request, Response, RequestBuilder> action, Request request, ActionListener<Response> listener) {
                     final Supplier<ThreadContext.StoredContext> supplier = threadContext.newRestorableContext(false);
                     try (ThreadContext.StoredContext ignore = stashWithHeaders(threadContext, filteredHeaders)) {
                         super.doExecute(action, request, new ContextPreservingActionListener<>(supplier, listener));
```

**Review by jasontedor — APPROVED** [review]
The approach looks good. Please wait for final approval from @jaymode though.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTask.java
Comment (jaymode) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTask.java]: wrap in a unmodifiable map?

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTask.java
Comment (jaymode) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTask.java]: wrap in a unmodifiable map?

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTask.java
Comment (jaymode) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTask.java]: do you have equals/hashcode unit tests? If not can you add some and ensure the headers aspect is appropriately tested?

File: x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskTests.java
Comment (jaymode) [inline x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskTests.java]: looks like this isn't used?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilege.java
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilege.java]: can you add a unit test in `PrivilegeTests`?

**Review by jaymode — CHANGES_REQUESTED** [review]
I left some requests for unit tests and minor changes

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTask.java
Comment (martijnvg) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTask.java]: This test: https://github.com/martijnvg/elasticsearch/blob/75ea05434003c2535c3b7e3e796dbd292fda1062/x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskTests.java#L17

also tests equals and hashcode, because it indirectly extends from `AbstractWireTestCase`.

**Comment by martijnvg** [comment]
@jaymode @jasontedor I've updated this PR.

**Review by jaymode — APPROVED** [review]
LGTM

**Pull request merged**
